src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,432 @@
+        frequency = FrequencyMhz,
+        datarate = Datarate,
+        snr = SNR,
+        region = Region,
+        gateway = Gateway
+    },
+    #config_route_v1_pb{
+        oui = OUI,
+        net_id = NetID
+    }
+) ->
+    packet_router_pb:encode_msg(
+        #packet_router_packet_report_v1_pb{
+            gateway_timestamp_ms = GatewayTimestamp,
+            oui = OUI,
+            net_id = NetID,
+            rssi = RSSI,
+            frequency = FrequencyMhz,
+            datarate = Datarate,
+            snr = SNR,
+            region = Region,
+            gateway = Gateway,
+            payload_hash = crypto:hash(sha256, Payload)
+        },
+        packet_router_packet_report_v1_pb
+    ).
+
+-spec encode_packet_size(EncodedPacket :: binary()) -> PacketSize :: binary().
+encode_packet_size(EncodedPacket) ->
+    PacketSize = size(EncodedPacket),
+    EncodedValue = binary:encode_unsigned(PacketSize),
+    pad_u32(EncodedValue).
+
+-spec pad_u32(binary()) -> binary().
+pad_u32(Bin) ->
+    <<0:((4 - size(Bin)) * 8), Bin/binary>>.
+
+-spec generate_file_name(WriteDir :: string()) -> FilePath :: string().
+generate_file_name(WriteDir) ->
+    Timestamp = erlang:system_time(millisecond),
+    FileName = "packetreport." ++ integer_to_list(Timestamp),
+    filename:join(WriteDir, FileName).
+
+-spec open_tmp_file(FilePath :: string(), WriteOptions :: [atom()]) ->
+    {ok, IODevice :: file:io_device()} | {error, Reason :: atom()}.
+open_tmp_file(FilePath, WriteOptions) ->
+    case file:open(FilePath, WriteOptions) of
+        {error, Error} ->
+            lager:error("failed to open tmp write file: ~p", [Error]);
+        IODevice ->
+            IODevice
+    end.
+
+-spec upload_file(
+    AWSClient :: aws_client:aws_client(),
+    FilePath :: binary(),
+    S3FileName :: binary(),
+    BucketName :: binary()
+) -> {ok, Response :: term()} | {error, upload_failed}.
+upload_file(AWSClient, FilePath, S3FileName, BucketName) ->
+    {ok, Content} = file:read_file(FilePath),
+
+    case
+        aws_s3:put_object(
+            AWSClient,
+            BucketName,
+            S3FileName,
+            #{
+                <<"Body">> => Content
+            },
+            ?AWS_RETRY_OPTIONS
+        )
+    of
+        {ok, _, Response} ->
+            {ok, Response};
+        Error ->
+            lager:error(
+                "failed to upload packet report: file: ~p, error: ~p", [FilePath, Error]
+            ),
+            {error, upload_failed}
+    end.
+
+-spec upload_window_elapsed(StartTime :: timestamp_ms(), UploadWindow :: timestamp_ms()) ->
+    WindowElapsed :: boolean().
+upload_window_elapsed(StartTime, UploadWindow) ->
+    Timestamp = erlang:system_time(millisecond),
+    Timestamp - StartTime >= UploadWindow.
+
+-spec get_packets_by_timestamp(Timestamp :: timestamp_ms()) -> [term()].
+get_packets_by_timestamp(Timestamp) ->
+    ets:select(?ETS, [{{'$1', '$2'}, [{'=<', '$1', Timestamp}], ['$2']}]).
```

Comment:
Since the table is a `bag`, I suspect that the rows will not be returned in timestamp order. Is that acceptable? If not, you'd have to figure out a way of making the key values unique and use `ordered_set` or sort them after the `ets:select`.

Reply:
Do we expect this `select` to return a huge number of rows? Should we limit the return count with `select/3` and then iterate on the continuation?

Reply:
Good points, taking a look at my ETS usage.

Reply:
Made the suggested changes:
1. Switched to an `ordered_set`, incrementing a counter to act as index in ETS.
2. Adjusted select patterns to use `<` to avoid the possible race condition.
3. Rewrote to write to the file in batches using `select/3`